[feature] Fix title description category inconsistency

zkraime opened the pull request:
<b>Purpose:</b>
To fix issue #2686 - project title and description could only be edited in the dashboard page as inline editables. 
And to replace PR #3093
<b>Changes:</b>
Title and description can now be edited in the project settings page as normal text and textarea inputs,
as shown: 
1-
![screen shot 2015-06-11 at 11 53 27 pm](https://cloud.githubusercontent.com/assets/11323460/8123353/06542df4-1096-11e5-9157-78ccf6eebf36.png)
2-
![screen shot 2015-06-11 at 11 53 50 pm](https://cloud.githubusercontent.com/assets/11323460/8123355/0654a9e6-1096-11e5-9bc6-d7bf7e544367.png)
3-
![screen shot 2015-06-11 at 11 54 10 pm](https://cloud.githubusercontent.com/assets/11323460/8123356/06577bd0-1096-11e5-8ce8-3200456b64f8.png)
4-
![screen shot 2015-06-11 at 11 54 42 pm](https://cloud.githubusercontent.com/assets/11323460/8123354/06547656-1096-11e5-980a-b1b06d3e083d.png)

<b>Side effects:</b>
One thing I noticed was that if you enter html as input for title, it saves and the title appears blank:


![screen shot 2015-06-12 at 12 03 20 am](https://cloud.githubusercontent.com/assets/11323460/8123446/7f23a4ac-1097-11e5-8dec-9733935df9d6.png)
![screen shot 2015-06-12 at 12 03 35 am](https://cloud.githubusercontent.com/assets/11323460/8123447/80b5527a-1097-11e5-8d11-65ea3ba8bb8c.png)


The validation for title probably needs to be refactored anyway. I wanted to use changeMessage.js for the title and description alert messages but wasn't sure how to do so since it seems like the messages can only bind to one element (category). So, I just made it display an error message in case of a failure:
![screen shot 2015-06-12 at 12 13 17 am](https://cloud.githubusercontent.com/assets/11323460/8123463/e4b8b6fe-1097-11e5-8c59-983d65863cc1.png)

sloria commented:
There should only be one submit button for both the title and the description.

sloria commented:
You should be able to use the ChangeMessageMixin; all it does is add a `message` and `messageClass` observable which are meant to be bound to a help text element.

sloria commented:
Also, be careful to use the correct CSS classes for validation errors--the red in your screenshot looks different from elsewhere on the site.

lyndsysimon commented:
Bug: special characters are not displayed properly in the title edit input:

![image](https://cloud.githubusercontent.com/assets/131414/8194138/976b94f8-1447-11e5-8445-493229af350c.png)

This project's title is "Alpha & Test"

lyndsysimon commented on website/templates/project/settings.mako:
```diff
@@ -68,7 +68,27 @@
+                    <input class="form-control" type="text" maxlength="200" data-bind="value: title, 
+                                                                            event: { keyup: dirtyTitleDescription }">
+                    <span id="title-input-message"></span>        
+                    <br>
```
`<br>` shouldn't be used for layout. Wrapping with a `div.form-group` is probably what you want.

See Bootstrap's documentation for forms: http://getbootstrap.com/css/#forms-example

lyndsysimon commented on website/templates/project/settings.mako:
```diff
@@ -68,7 +68,27 @@
+                   
+                    <label for="title">Title</label>
+                    <input class="form-control" type="text" maxlength="200" data-bind="value: title, 
+                                                                            event: { keyup: dirtyTitleDescription }">
```
Don't bind to an event - use Knockout's `computed` or `pureComputed`.

Knockout docs: http://knockoutjs.com/documentation/computedObservables.html

lyndsysimon commented on website/static/js/projectSettings.js:
```diff
@@ -59,6 +73,17 @@ var NodeCategorySettings = oop.extend(
+            document.getElementById('title-input-message').innerHTML = 'Title cannot be blank.';
+        },
+        updateDescriptionError: function() {
+            document.getElementById('description-input-message').style.color = '#BD362F';
```
You needn't use `document.getElementById` - jQuery is available. Likewise, don't use the `.style.color`; use jQuery's methods instead.

This line would be:

   $('description-input-message').css({color = '#BD362F'})

That said, you also shouldn't be setting the color manually :) Use Bootstrap's classes - for an error, you might use `text-danger`.

See how other form validations on the OSF are done, especially those which use an HTML element that is bound to a property of the viewmodel.

lyndsysimon commented:
I'm not quite ready to mark this as complete for first pass. Please make the changes above and we'll go from there.

zkraime commented:
@lyndsysimon Thanks, yes I agree. I'll go back over this and make the corrections

zkraime commented:
@lyndsysimon Can this get a second look? thanks!

lyndsysimon commented on website/static/js/projectSettings.js:
```diff
@@ -38,19 +44,38 @@ var NodeCategorySettings = oop.extend(
             self.dirty = ko.observable(false);
+
+            self.dirtyTitleDescription = ko.computed(function(){
+                if (self.title() !== self.decodedTitle || 
```
This if/else block can be simplified as:

```javascript
return (self.title() !== self.decodedTitle || self.description() !== self.decodedDescription)
```

lyndsysimon commented on website/static/js/projectSettings.js:
```diff
@@ -38,19 +44,38 @@ var NodeCategorySettings = oop.extend(
                     self.dirty(true);
                 }
-            });
+            });    
```
Please remove the whitespace at the end of this line.

lyndsysimon commented on website/static/js/projectSettings.js:
```diff
@@ -38,19 +44,38 @@ var NodeCategorySettings = oop.extend(
-        updateError: function(xhr, status, error) {
+        //TODO - refactor title/description success handler to display success message 
+        updateTitleDescriptionSuccess: function() {
+            window.location.reload();   
```
Please remove the trailing whitespace here as well.

Also, why do we have to reload the page?

zkraime replied:
We shouldn't need to - I was going to just display a success message but wasn't sure if that's best.
Thanks for pointing every out, I'll work on them

lyndsysimon replied:
After further review, I'm OK with reloading the page. It's not perfect, but updating the title outside of Knockout or leaving it displayed incorrectly are both greater evils :)

zkraime replied:
Roger that

lyndsysimon commented on website/static/js/projectSettings.js:
```diff
@@ -67,19 +101,45 @@ var NodeCategorySettings = oop.extend(
+            return $osf.putJSON(self.titleDescriptionEditUrl, {
+                    title: $osf.htmlEscape(self.title()),
+                })
+                .done(self.updateDescription.bind(self))
```
It looks like the functions here are swapped - shouldn't this be `self.updateTitle.bind(self)`?

zkraime replied:
@lyndsysimon The success handler for updating just the category used to be named 'updateSuccess' but I changed it to specify/distinguish it from the title/description handlers

lyndsysimon replied:
Right - what I'm not understanding is why you're calling `self.updateDescription` inside `updateTitle`. Either the wrong function is getting called, or the success funciton should be renamed to show that it gets called after updating more than just the description.

zkraime replied:
@lyndsysimon Right ok. After Steve clarified that we should use one button to save changes to both title and description, I originally I had one update handler that made two ajax requests but it didn't work properly. Therefore, I made it so that the title update handler gets called first, and if the request is successful, it then goes to the description update handler which has its own ajax request. If that completes without error as well, there's one corresponding title/description success handler which only reloads the window now, but that I will change to output a success message.

zkraime commented:
@lyndsysimon I'm sorry, after I pulled from osf, I stashed before committing and switched branches. When I switched back and applied the stash it considered a bunch of files as changed. Should I create a new PR?

lyndsysimon commented:
You shouldn't need to create a new PR. If I understand what you did in git, the changes should be your changes, so committing them and pushing them would be the right thing to do.

If the changes aren't your changes, it's possible that you stashed a merge in-progress, in which case I'd have to see more to suggest the next step.

zkraime commented:
@lyndsysimon I just noticed that it now says 213 files changed, most of which werent my own changes so I would venture to say the latter happened. Figured it would just be simpler to redo the PR, since the images needed to be updated anyway

zkraime commented:
@sloria Would it be helpful if I resubmitted this PR?

sloria commented:
@zkraime Yes please; I'm not sure how all those other changes got into this PR.

Thanks!

zkraime commented:
Replaced by PR #3390
